Switch default handler assignment to skip locked or transaction isolation if possible

mvdbeek opened the pull request:
## What did you do? 
- Switch default handler assignment mechanism to db skip locked or transaction isolation, depending on what the database connection supports
- Add the option to use DB-TRANSACTION-ISOLATION if the database doesn't support RETURNING (sqlite before 3.35.0)

## Why did you make this change?
Our recommended production database (postgres) has been supporting the necessary features (the RETURNING statement plus isolation level SERIALIZABLE ) since at least version 8. sqlite has recently added the RETURNING statement (and we don't necessarily need the returning statement ...).

This will help with deprecating uwsgi mules and should be the most robust mechanism and should also be the easiest way for re-assigning running jobs when handlers go away (which is to set jobs / invocations to the default tag) or when adding additional handlers.
This is also the easiest way to have celery tasks submit jobs (which can't use the default DB-SELF / MEM-SELF mechanism).

## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

## For UI Components
- [ ] I've included a screenshot of the changes

mvdbeek committed:
Change default handler assignment to db-skip-locked

mvdbeek committed:
Figure out which assignment method is best given the deployed database

nsoranzo commented on lib/galaxy/web_stack/handlers.py:
```diff
@@ -53,6 +54,33 @@ def add_handler(self, handler_id, tags):
+        try:
+            model.session.execute(stmt)
+            return True
+        except Exception:
```
Is there a more specific exception we can catch here and in `supports_skip_locked()` ?

Also, we could `log.debug()` a message.

mvdbeek replied:
This would be hypothetical, since sqlalchemy forwards the underlying library exception, so would be different for psycopg2 / asyncpg / sqlite3. I'll think about what we can log here.

mvdbeek replied:
fwiw these are the exception I've seen so far:
For psycopg2 + postgres 8 and the exception for with_for_update(skip_locked=True) is 

```
ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "SKIP"
LINE 3: WHERE job.id =  -1 FOR UPDATE SKIP LOCKED
```

for sqlite (sqlalchemy hasn't updated this yet, since the first release that supports it is very new) this is currently
```
CompileError: RETURNING is not supported by this dialect's statement compiler.
```

If I remove that explicit exception from sqlalchemy and replace it with the actual implementation it either works (sqlite >= 3.35.0) or fails with
```
OperationalError: (sqlite3.OperationalError) near "RETURNING": syntax error
[SQL: UPDATE job SET update_time=? WHERE job.id = ? RETURNING job.id]
[parameters: ('2021-04-06 11:33:26.931462', -1)]
(Background on this error at: http://sqlalche.me/e/13/e3q8)
```

nsoranzo replied:
Ah, I thought we would get a SQLAlchemy error. Then I guess a log message would be good enough.

mvdbeek committed:
Adjust unit tests for new setup

mvdbeek committed:
Add log statements when determining best handler assignment method

mvdbeek committed:
Update documentation for new defaults and recommendations

mvdbeek committed:
Add generic GALAXY_WEB_PORT variable to infrastructure_url parsing

mvdbeek committed:
Run kubernetes staging test against uvicorn

mvdbeek committed:
Fix handlers pools, job handler capacilities

mvdbeek committed:
Fix workflow_handler_configuration tests

mvdbeek committed:
Fix log statement

mvdbeek committed:
Setup grabber query after fork

mvdbeek committed:
Shut down pulsar app on Galaxy shutdown if embedded

Fixes occasional error in pulsar embedded integration tests,
which is a neverending stream of:
```
2021-04-12T08:09:18.5912970Z FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpk0dfc3gg/tmp539e3wrp/tmp3zkj6xnj/database/pulsar_persisted_data/_default_-active-jobs'
2021-04-12T08:09:18.5916534Z pulsar.managers.stateful ERROR 2021-04-12 08:09:18,347 [pN:main,p:3088,tN:[manager=_default_]-[action=monitor]] Failure in stateful manager monitor step.
2021-04-12T08:09:18.5917629Z Traceback (most recent call last):
2021-04-12T08:09:18.5919687Z   File "/home/runner/work/galaxy/galaxy/galaxy root/.venv/lib/python3.7/site-packages/pulsar/managers/stateful.py", line 361, in _run
2021-04-12T08:09:18.5920924Z     self._monitor_active_jobs()
2021-04-12T08:09:18.5924333Z   File "/home/runner/work/galaxy/galaxy/galaxy root/.venv/lib/python3.7/site-packages/pulsar/managers/stateful.py", line 366, in _monitor_active_jobs
2021-04-12T08:09:18.5925618Z     active_job_ids = self.stateful_manager.active_jobs.active_job_ids()
2021-04-12T08:09:18.5927558Z   File "/home/runner/work/galaxy/galaxy/galaxy root/.venv/lib/python3.7/site-packages/pulsar/managers/stateful.py", line 307, in active_job_ids
2021-04-12T08:09:18.5929117Z     job_ids = os.listdir(target_directory)
```

mvdbeek committed:
Workaround missing RETURNING for transaction isolation

In that case there just won't be any logs about which jobs were grabbed,
but that's probably justified by being able to drop db-self and
db-preassgin which would have no real reason to remain options.

mvdbeek committed:
Remove db-self, db-preassign fallback options

mvdbeek committed:
Move get_preferred_handler_assignment_method into ApplicationStack

and simplify _set_default_job_handler_assignment_methods

mvdbeek committed:
Restore stack pool default setup

mvdbeek committed:
Make sure web handlers don't become job handlers if handlers section exists

And add a default pool, so that exisiting job setups that don't specify
a pool (for instance those that previously used db-preassign) will just
continue working.

jmchilton commented on test/integration/test_kubernetes_staging.py:
```diff
@@ -208,7 +206,7 @@ def test_mulled_simple(self):
 
 def set_infrastucture_url(config):
-    infrastructure_url = "http://%s:$UWSGI_PORT" % GALAXY_TEST_KUBERNETES_INFRASTRUCTURE_HOST
+    infrastructure_url = "http://%s:$GALAXY_WEB_PORT" % GALAXY_TEST_KUBERNETES_INFRASTRUCTURE_HOST
```
Awesome - thanks!
